From 39af483bf626f062b28de32e96e7735b723d22f6 Mon Sep 17 00:00:00 2001
From: Bharat Bhushan <Bharat.Bhushan@nxp.com>
Date: Mon, 10 Dec 2018 14:53:33 +0530
Subject: [PATCH 397/741] vfio/fsl-mc: corrected device removal sequence

commit 0e1ebba44c14e5e1c7e328b42e7b46cb8be717d9 from
http://source.codeaurora.org/external/qoriq/qoriq-components/linux

We need to free allocated devices on behalf of dprc
before initiating the dprc device itself.

Signed-off-by: Bharat Bhushan <Bharat.Bhushan@nxp.com>
Signed-off-by: Zhantao Tang <zhantao.tang@windriver.com>
---
 drivers/vfio/fsl-mc/vfio_fsl_mc.c | 13 +++++++++++--
 1 file changed, 11 insertions(+), 2 deletions(-)

diff --git a/drivers/vfio/fsl-mc/vfio_fsl_mc.c b/drivers/vfio/fsl-mc/vfio_fsl_mc.c
index 10738a70f0ae..210c98041a05 100644
--- a/drivers/vfio/fsl-mc/vfio_fsl_mc.c
+++ b/drivers/vfio/fsl-mc/vfio_fsl_mc.c
@@ -618,6 +618,13 @@ static int vfio_fsl_mc_device_remove(struct device *dev, void *data)
 	if (WARN_ON(mc_dev == NULL))
 		return -ENODEV;
 
+	if (strcmp(mc_dev->obj_desc.type, "dprc") == 0) {
+		device_for_each_child(&mc_dev->dev, NULL, vfio_fsl_mc_device_remove);
+		dprc_close(mc_dev->mc_io, 0, mc_dev->mc_handle);
+		fsl_mc_portal_free(mc_dev->mc_io);
+		mc_dev->mc_io = NULL;
+	}
+
 	fsl_mc_device_remove(mc_dev);
 	return 0;
 }
@@ -640,8 +647,10 @@ static void vfio_fsl_mc_cleanup_dprc(struct vfio_fsl_mc_device *vdev)
 	dev_set_msi_domain(&mc_dev->dev, NULL);
 
 	fsl_mc_cleanup_all_resource_pools(mc_dev);
-	dprc_close(mc_dev->mc_io, 0, mc_dev->mc_handle);
-	fsl_mc_portal_free(mc_dev->mc_io);
+	if (mc_dev->mc_io) {
+		dprc_close(mc_dev->mc_io, 0, mc_dev->mc_handle);
+		fsl_mc_portal_free(mc_dev->mc_io);
+	}
 }
 
 static int vfio_fsl_mc_probe(struct fsl_mc_device *mc_dev)
-- 
2.17.1

